### PR TITLE
Cache prompt log probabilities with prefix blocks

### PR DESCRIPTION
**Pull request by lmcafee-nvidia**
## What does this PR do?

This PR lets prefix caching reuse prompt log probabilities when the prompt and score settings match, and preserves original-prompt scores across cache-disabled recompute checkpoints.

- It stores selected prompt scores and requested top-N scores in compact, mutable sidecars on canonical KV blocks.
- It reuses scores only when every setting that can affect them matches.
- It fills missing rows without replacing valid cached rows during hybrid-model replay.
- It carries chunk-boundary scores on the request until the continuation is admitted.
- It reconstructs complete prompt-score results for completion and streaming without private KV copies or a fallback scheduler.
- It keeps canonical and deprecated prompt-score controls synchronized when a sidecar-backed request is checkpointed for recomputation.
- It promotes the first complete cache-disabled prompt-score recomputation back to the original record segment before merging or coordinator streaming.
- It supports GPT and hybrid models under legacy and asynchronous scheduling.
- It makes the chat endpoint return generated-token scores from the correct request field while retaining the current finite-value sanitizer.

## Example failure scenario(s)

**Scenario 1 (matching scores)**

- Request A has 513 prompt tokens and asks for top-5 scores.
- Request A caches 2 full blocks and their reusable prompt scores.
- Request B repeats the 513 tokens with the same settings.
- Request B reuses the canonical blocks and stored scores while computing only the required tail.
- Request B returns the same complete result as cache-disabled inference.

**Scenario 2 (different settings)**

- Request A caches top-5 scores for a 513-token prompt.
- Request B repeats the prompt but asks for top-4 scores.
- The exact sidecar key rejects Request A's top-5 values.
- Request B recomputes top-4 scores on the canonical KV blocks.
- Request B returns correct rows without receiving false cache-hit credit.

**Scenario 3 (chunk boundary)**

- Request A has 513 prompt tokens and reaches a prompt-chunk boundary.
- The first chunk produces the score for the next prompt token.
- Request A carries that score until its continuation is admitted.
- The next chunk attaches it to the correct sidecar row.
- Legacy and asynchronous scheduling return the same complete score sequence.

**Scenario 4 (hybrid replay)**

- Request A caches top-5 scores for 2 full prefix blocks.
- Request B restores recurrent state after the first block.
- Request B replays the second block to continue the prompt.
- Replay preserves valid cached rows and fills only missing rows.
- Request B matches the cache-disabled result.

**Scenario 5 (memory pressure)**

- Request A leaves 2 matching canonical prefix blocks pinned.
- Only 1 KV block remains allocatable.
- Request B repeats the prompt and needs 1 tail block.
- Request B reuses the 2 pinned blocks and finishes.
- No shadow blocks or fallback scheduling path are needed.

**Scenario 6 (HTTP generated scores)**

- Request A generates 2 tokens with scores enabled.
- The generated scores are `-0.25` and `-0.50`.
- A legacy field contains `-9.0` and `-9.0`.
- The endpoint returns `-0.25` and `-0.50`.
- A request that opts out still receives no scores.

**Scenario 7 (checkpointed top-N scores)**

- Request A reuses 511 of its 513 prompt tokens and asks for top-5 scores.
- The engine checkpoints Request A after its first generated token.
- The sidecars first materialize all 512 original prompt-score rows.
- The checkpoint segment skips recalculating those prompt scores.
- Request A resumes and returns the same complete scores as cache-disabled inference.

**Scenario 8 (cache-disabled recompute)**

- Request A asks for prompt and top-3 scores with prefix caching disabled.
- The async engine starts A but has not yet attached its pending prompt scores.
- The engine suspends and checkpoints A for recomputation.
- A later segment recomputes the complete prompt-score rows and restores them to the first segment.
- A returns the same complete prompt scores as legacy recompute.

## How was this bug discovered?

A cache-off versus cache-on comparison first showed that generated tokens matched while the cache-on result omitted prompt scores skipped by prefix reuse. The oracle was the complete cache-disabled token, selected-score, and top-N result.

Focused regressions then crossed exact and mismatched score settings with real chunk boundaries, concurrent variants, one-free-block pressure, allocator cleanup, GPT attention, hybrid recurrent-state restoration, and both scheduling modes. Exact hybrid comparison exposed that replay could replace already-valid cached rows with slightly different values, so valid rows are now authoritative and only missing rows are filled.

The HTTP defect was exposed by placing different sentinel values in the legacy and generated score fields. The response returned the legacy values instead of the engine's generated-token values.

Reconciliation with fork PR #50 added a real prefix-hit checkpoint comparison. It showed that score materialization was correct, but the checkpoint copied a stale deprecated top-N control that contradicted the new skip setting and stopped the request before resume.

Adding prompt and top-3 scores to the existing asynchronous RECOMPUTE suspend/resume pair exposed a separate cache-disabled defect. HSG job `7003965` reproduced it: legacy returned 8 prompt rows while async returned 0 because a later checkpoint segment held the complete recomputation while result publication still read the empty first segment. The fix promotes only the earliest complete zero-based recomputation and leaves prefix-offset reconstruction with #7063's position-aware sidecars.

## Dependency and scope

The original [fork PR #60](https://github.com/lmcafee-nvidia/Megatron-LM/pull/60) depended on the prefix-cache foundation in NVIDIA/Megatron-LM #6418. That foundation and its prerequisites are now in `main` through merged [NVIDIA/Megatron-LM #7031](https://github.com/NVIDIA/Megatron-LM/pull/7031), whose squash commit is `6807fb8e45c634e933bfd861e889486fd0167be3`.

This upstream landing PR was originally ported from `main` at `1e7598cbfae888cdd3d741a351aae588d56f66c0` and was reconciled through current `main` at `c9b53d0a87cb926f47115259593ecfeb351ca29f` by signed merge `5b3de84618dc88c6fbfaacfd03fee3a6f450b5e5`; fork PR #60 remains unchanged as the historical review record. The port preserves current-main post-admission cache-hit accounting, disaggregation handoff APIs, streaming behavior, and chat-response sanitization.

A sidecar key includes every setting that can change a stored score. Each physical KV block exposes at most one current sidecar variant, active requests retain references to variants they already consumed, and sidecars follow ordinary KV-block cleanup and eviction. Cached `top_n_logprobs` is limited to 100, keeping a 256-token sidecar below 1 MiB.

The diff changes 7 production files and 7 focused unit-test files. It adds 348 net unit-test lines, bringing the prefix-cache series represented by #7031 plus this port to 1,388 net lines, 612 below the 2,000-line budget. It changes no functional tests or recipes.

## Validation

- Exact tested commit: `a928ea1754e63c28fb14759da4f9cd1c4f0cafdc`.
- DFW H100 job `18094042` tested predecessor `5b3de84618dc88c6fbfaacfd03fee3a6f450b5e5`: ranks 0 through 7 each passed all 117 selected tests with zero failures, zero skips, and zero deselections.
- HSG job `7003552` tested the predecessor production tree represented by `6502eeccb2bbb90d8fb8e2b234b6e3da376523bf` with two assessment-only real checkpoint regressions: all 4 ranks passed the cache-disabled and 511-token prefix-hit recompute cases with zero failures and zero skips.
- HSG job `7003685` tested predecessor `6502eeccb2bbb90d8fb8e2b234b6e3da376523bf`: all 4 ranks passed the GPT and hybrid prompt-score sidecar matrix under legacy and asynchronous scheduling, with zero failures and zero skips.
- HSG job `7004172` tested the exact head: all 4 ranks passed all 6 selected tests covering cache-disabled async RECOMPUTE prompt-score parity, per-request score policy, and the GPT/hybrid sidecar matrix under legacy and asynchronous scheduling, with zero failures and zero skips.
- The fused-RoPE stress row ran and passed on every DFW rank without a retry.
- Exact-key reuse, mismatched-key rejection, mixed valid/missing replay, memory pressure, concurrent variants, streaming materialization, and handoff ordering all passed on every DFW rank.
- `git diff --check`, isort, Black, Ruff, and Python byte-compilation checks passed on the exact tip.

### CODE REVIEW

File: megatron/core/inference/contexts/dynamic_context.py [hunk 1]
```diff
@@ -3169,7 +3222,13 @@ def _compute_prefix_match(
...
                 num_mamba_matched <= num_matched
             ), f"Mamba match ({num_mamba_matched}) > KV match ({num_matched})"
-            if num_mamba_matched > 0 and block_aligned:
+            if prompt_logprob_key is not None and block_aligned:
+                # Recurrent state is stored only at block boundaries. Restore
+                # the preceding boundary and recompute the final sidecar-backed
+                # block so the first uncached prompt score has the right state.
+                executable_blocks = min(num_mamba_matched, num_logprob_matched)
+                prefix_skip_tokens = max(0, executable_blocks - 1) * self.block_size_tokens
+            elif num_mamba_matched > 0 and block_aligned:
```
Comment (tdene) [inline megatron/core/inference/contexts/dynamic_context.py:3231]: There is a bug. I think the best way of explaining is providing this diff to your unit-test that exposes it:

```
diff --git a/tests/unit_tests/inference/contexts/test_dynamic_prefix_caching.py b/tests/unit_tests/inference/contexts/test_dynamic_prefix_caching.py
index 7dadea68f1..2a1c5c2ee4 100644
--- a/tests/unit_tests/inference/contexts/test_dynamic_prefix_caching.py
+++ b/tests/unit_tests/inference/contexts/test_dynamic_prefix_caching.py
@@ -3218,12 +3218,13 @@ class TestPrefixCacheRealEngineMatrix(DynamicInferenceEngineTestBase):
             cache_config.materialize_only_last_token_logits = False
             cache_config.enable_chunked_prefill = True
             cache_config.async_sched_mode = async_sched_mode
-            cache_config.context_max_tokens = cache_config.context_block_size_tokens + 8
+            if model_provider == "gpt":
+                cache_config.context_max_tokens = cache_config.context_block_size_tokens + 8
             cache_env = self._build_test_env(cache_config)
             engine = cache_env.engine
             engine.controller.tokenizer.detokenize = lambda tokens, **_: f"tok_{tokens[0]}"
             donor, donor_cost, donor_chunked = run(engine, 10, prompt, 5)
-            assert donor_cost == prompt_length and donor_chunked
+            assert donor_cost == prompt_length and (donor_chunked or model_provider == "hybrid")
             self._assert_prompt_logprob_parity(donor, oracle_top5)
             allocator = engine.context.kv_block_allocator
             sidecars = sorted(
```

With this diff, we go on to later fail inside `self._assert_prompt_logprob_parity(exact, oracle_top5)`. I'm just changing the assert in this diff so that we don't erroneously fail early.

File: tests/unit_tests/inference/contexts/test_dynamic_prefix_caching.py [hunk 2]
```diff
@@ -3136,6 +3135,163 @@ def _assert_finite_completion(request):
...
+        for field in ("prompt_top_n_logprobs",):
+            actual_rows, expected_rows = getattr(actual, field), getattr(expected, field)
+            assert actual_rows is not None and expected_rows is not None
+            assert len(actual_rows) == len(expected_rows)
+            for actual_row, expected_row in zip(actual_rows, expected_rows):
+                cls._assert_top_n_logprobs_close(actual_row, expected_row)
+
+    @pytest.mark.internal
+    @pytest.mark.parametrize("model_provider", ["gpt", "hybrid"])
+    @pytest.mark.parametrize("async_sched_mode", list(AsyncScheduleMode))
```
Comment (tdene) [inline tests/unit_tests/inference/contexts/test_dynamic_prefix_caching.py:3187]: This `async_sched_mode` parameter doesn't actually get fed into `InferenceConfig` inside `TestPrefixCacheRealEngineMatrix._build_inference_context`?

File: megatron/core/inference/contexts/dynamic_context.py (hunk 1 above)
Comment (tdene) [reply megatron/core/inference/contexts/dynamic_context.py:3231]: The problem happens when the "donor" is not chunked.

`compute_and_store_offsets` decides when to save snapshots. For a 513-token unchunked donor, `kv_div_abs` doesn't match anything, and `last_aligned_abs` and `penultimate_abs` both match token 512 (end of block 1). If the donor was chunked, the chunk-end clause would save a snapshot at token 256, but that doesn't happen for unchunked.

The next request arrives with the same prompt and matches. After this PR, that request tries to get the penultimate snapshot. The penultimate snapshot doesn't exist; the request is trying to resume from an undefined `matched_block_ids[executable_blocks − 2]`. There is no check on whether that snapshot was actually made, we just try to load it regardless.

The `matched_block_ids[executable_blocks − 2]` trace is `prefix_skip_tokens = max(0, executable_blocks - 1) * self.block_size_tokens` -> `restore_block_count = prefix_skip_tokens // self.block_size_tokens` -> `matched_block_ids[restore_block_count − 1]`.

File: tests/unit_tests/inference/contexts/test_dynamic_prefix_caching.py (hunk 2 above)
Comment (lmcafee-nvidia) [reply tests/unit_tests/inference/contexts/test_dynamic_prefix_caching.py:3226]: You're right, that setting wasn't reaching `InferenceConfig`. It now gets passed through, and the test checks that the selected scheduler actually runs. Both modes pass against the same cache-disabled reference; removing the fix makes the legacy cases fail.

File: megatron/core/inference/contexts/dynamic_context.py (hunk 1 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/contexts/dynamic_context.py]: The fix now finds a usable saved state instead of assuming the previous block has one. I kept your unchunked case and added the chunked case back: the unchunked request recomputes all 513 tokens, while the chunked request restores at token 256. Both match cache-disabled results under both schedulers, and undoing the fix brings the bug back.

**Review by tdene — APPROVED** [review]

**Review by shanmugamr1992 — APPROVED** [review]
